clang/test/CodeGenSYCL/code_location.cpp

```diff
@@ -0,0 +1,266 @@
+  return 0;
+}
+// CHECK: template <> struct KernelInfo<::KernelName3<::KernelName2>> {
+// CHECK:   static constexpr const char* getFileName() {
+// CHECK: #ifndef NDEBUG
+// CHECK:     return "code_location.cpp";
+// CHECK: #else
+// CHECK:     return "";
+// CHECK: #endif
+// CHECK:   }
+// CHECK:   static constexpr const char* getFunctionName() {
+// CHECK: #ifndef NDEBUG
+// CHECK:     return "::KernelName3<::KernelName2>";
+// CHECK: #else
+// CHECK:     return "";
+// CHECK: #endif
+// CHECK:   }
+// CHECK:   static constexpr unsigned getLineNumber() {
+// CHECK: #ifndef NDEBUG
+// CHECK:     return 125;
+// CHECK: #else
+// CHECK:     return 0;
+// CHECK: #endif
+// CHECK:   }
+// CHECK:   static constexpr unsigned getColumnNumber() {
+// CHECK: #ifndef NDEBUG
+// CHECK:     return 44;
+// CHECK: #else
+// CHECK:     return 0;
+// CHECK: #endif
+// CHECK:   }
+// CHECK: };
+
+// Check that the location information returned is that of l4
+auto l4 = []() { return 4; };
+int test4() {
+  cl::sycl::queue q;
+  q.submit([=](cl::sycl::handler &h) { h.single_task<class KernelName4>(l4); });
+  return 0;
+}
+// CHECK: template <> struct KernelInfo<KernelName4> {
+// CHECK:   static constexpr const char* getFileName() {
+// CHECK: #ifndef NDEBUG
+// CHECK:     return "code_location.cpp";
+// CHECK: #else
+// CHECK:     return "";
+// CHECK: #endif
+// CHECK:   }
+// CHECK:   static constexpr const char* getFunctionName() {
+// CHECK: #ifndef NDEBUG
+// CHECK:     return "KernelName4";
+// CHECK: #else
+// CHECK:     return "";
+// CHECK: #endif
+// CHECK:   }
+// CHECK:   static constexpr unsigned getLineNumber() {
+// CHECK: #ifndef NDEBUG
+// CHECK:     return 160;
+// CHECK: #else
+// CHECK:     return 0;
+// CHECK: #endif
+// CHECK:   }
+// CHECK:   static constexpr unsigned getColumnNumber() {
+// CHECK: #ifndef NDEBUG
+// CHECK:     return 11;
+// CHECK: #else
+// CHECK:     return 0;
+// CHECK: #endif
+// CHECK:   }
+// CHECK: };
+
+// Check that fully qualified name is returned when unnamed lambda
+// kernel is enclosed in a namespace
+namespace NS {
+int test5() {
+  cl::sycl::queue q;
+  q.submit([=](cl::sycl::handler &h) { h.single_task([] {}); });
+  q.submit([=](cl::sycl::handler &h) { h.single_task<class KernelName5>([] {}); });
+  return 0;
+}
+} // namespace NS
+// CHECK: template <> struct KernelInfoData<'_', 'Z', 'T', 'S', 'Z', 'Z', 'N', '2', 'N', 'S', '5', 't', 'e', 's', 't', '5', 'E', 'v', 'E', 'N', 'K', 'U', 'l', 'R', 'N', '2', 'c', 'l', '4', 's', 'y', 'c', 'l', '7', 'h', 'a', 'n', 'd', 'l', 'e', 'r', 'E', 'E', '_', 'c', 'l', 'E', 'S', '3', '_', 'E', 'U', 'l', 'v', 'E', '_'> {
+// CHECK:   static constexpr const char* getFileName() {
+// CHECK: #ifndef NDEBUG
+// CHECK:     return "code_location.cpp";
+// CHECK: #else
+// CHECK:     return "";
+// CHECK: #endif
+// CHECK:   }
+// CHECK:   static constexpr const char* getFunctionName() {
+// CHECK: #ifndef NDEBUG
+// CHECK:     return "NS::";
```

Comment:
This seems wrong.

Reply:
Similar thoughts to https://github.com/intel/llvm/pull/5335/files/190a44de31b432c907c062ba9ffd4f09a36ff7db#r798059401 
Since there is an enclosing namespace, that's the only (unmangled) name related information we have for pinpointing this lambda, combined with its location, of course. 

Could you please share your thoughts? @alexbatashev

Reply:
Same as above. Naming unnamed kernels is tricky. I suggest leave it as-is and change it later if we feel so.